README.md
<img src="https://raw.githubusercontent.com/MicahElliott/wunderbar/master/img/wunderbar.jpg" align="left" hspace="5px"/>

[Configuration][2] | [Running][3] | [i3status][4] | [Shell Prompt][5]

Wunderbar is a "widget" for
i3bar/[i3status](https://github.com/i3/i3status) (or any text based
display like prompts or conky).

**Wunderbar prints a concise display of current and forecasted
weather.**

    Cl48/W3 | Cl48 MC45 MC44 MC45 | MC4357 CR5054 CR4054 CR4055 ...
    ^^^^^^^   ^^^^^^^^^^^^^^^^^^^   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    present       next 12 hrs               next 7 days

The above tells me I may not be spending much time outside for the
next several days.  More specifically, the present condition is
"Cloudy, 48 degrees, a little wind.  The next hours (4 indicators, 3
hours apart) and days will be "Mostly Cloudy" with "Chances of Rain".
See the source (`conds` table) for a little more detail on the weather
conditions mnemonics.


## Configure Wunderground and Wunderbar

Get your own free Wunderground API key
by [signing up](https://www.wunderground.com/signup?mode=api_signup).
Quick and easy (beware the auto-play audio ads!).  Tell Wunderbar
about your key (and other things) by exporting environment variables
in your shell:

    % export WUNDERKEY=abc123  # but use your own!

Find your approximate location
at [wunderground.com](https://www.wunderground.com/) by searching.
Then click _Change Station_ to get more specific about your
nearest
[personal weather station](https://www.wunderground.com/weatherstation/overview.asp) (PWS).
Alternatively, you can just use your zip code (or regional
equivalent).  An example for _Beaverton, OR, USA_ is:

    % export WUNDERLOC=KORBEAVE74  # a pws near me
    OR
    % export WUNDERLOC=97005

Set your _temperature units_ for Fahrenheit (`f` (default)) or Celsius
(`c`).

    % export WUNDERUNITS=f

You might want to display your PWS ID if you move around much.

    % export WUNDERSHOWLOC=yes


## Run it

Clone this repo.

Build an uberjar (generate something like
`target/wunder-0.1.0-SNAPSHOT-standalone.jar`):

    % lein uberjar

Put Weatherbar into your crontab to run every half-hour.

    % crontab -e
    ...
    */30 * * * *    WUNDERKEY=123... WUNDERLOC=456... java -jar /path/to/wunder-0.1.0-SNAPSHOT-standalone.jar

On each run, a temporary file, `/tmp/wunderbar.txt`, is created/replaced,
holding the one-line text of the current run.

While testing, be careful about making runs in quick succession.
The
[API's free plan](https://www.wunderground.com/weather/api/d/pricing.html?MR=1) wants
to limit you to 10 calls/minute, and each Wunderbar run costs you 3
calls.

If you want to see logging output:

    % export WUNDERDEBUG=yes

It should be enough to **update (run) every 15 or 30 minutes**.
Although it may be slow to run (a second or two), the status bar
itself will be very quick to update (e.g., you may update your actual
bar every second or few) since it's only reading a temporary file —
not calling anything.


## Add to i3status

Modify your `~/.i3/config` to call a custom `my-i3status` command (as
described
[here](https://i3wm.org/i3status/manpage.html#_external_scripts_programs_with_i3status).

    bar {
      ...
      status_command my-i3status
    }

Then your `my-i3status` script is something like:

    i3status --config ~/.i3status.conf | while :
    do
      read line
      weather=$(cat /tmp/wunderbar.txt)
      echo "$weather | $line" || exit 1
    done


## As a shell prompt

Instead of printing the whole long line, you could just _cut_ present
conditions (`Cl48/W3`) and put it into your shell prompt.  Here's all
you need for a simple Zsh prompt:

    % PROMPT='$(cut -f1 -d" " /tmp/wunderbar.txt) %F{red}%n%f@%F{blue}%m%f %F{yellow}%1~%f %# '
    Cl50/W0 mde@arix ~ %


## Improvement ideas

I couldn't figure out how to get color working, but that would be
awesome if someone knows how to feed this through i3status as json
with markup.  Highs in red, lows in blue, a few other colors would
make it more quickly readable.

A lunch-time weather report could be spoken to you.  Would be fun to
parse the concise data generated by Wunderbar into a mini weather
report.

Might be worth adding _wind_ reporting if you live in a windy city.
It's already captured, but I don't display it for brevity's sake.

You could have a randomly rotating weather status showing all the
places you'd rather be, like Hawaiʻi.

Could auto-detect location based on IP, etc, and find nearby PWS.


## Prior work

This project started out as a Zsh script using OpenWeatherMap.  I
found Wunderground to be a more reliable and accurate service and
wanted to rewrite it in Clojure.  So this builds on some things
learned from Weatherbar.  This version is missing OWM's sunup/sundown
feature, which was kinda' nice.


## License

Copyright © 2017 Micah Elliott

Distributed under the Eclipse Public License, the same as Clojure.


[2]: #configure-wunderground-and-wunderbar
[3]: #run-it
[4]: #add-to-i3status
[5]: #as-a-shell-prompt


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-04-26
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-04-26
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/wunder/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/wunder/compare/0.1.0...0.1.1


test/wunder/core_test.clj
(ns wunder.core-test
  (:require [clojure.test :refer :all]
            [wunder.core :refer :all]
            [clojure.data.json :as json]))

(def sfo (json/read-json (slurp "sfo-forecast.json")))

;; (select-present) =>
(def present-data
  {:windchill_c "NA",
   :relative_humidity "50%",
   :local_tz_long "America/Los_Angeles",
   :dewpoint_string "55 F (13 C)",
   :precip_1hr_string "-999.00 in ( 0 mm)",
   :forecast_url
   "http://www.wunderground.com/US/CA/Beverly_Hills.html",
   :local_time_rfc822 "Wed, 26 Apr 2017 10:38:56 -0700",
   :heat_index_c "NA",
   :temp_f 75.0,
   :feelslike_f "75.0",
   :pressure_in "29.90",
   :precip_today_in "0.00",
   :windchill_f "NA",
   :wind_degrees 270,
   :dewpoint_c 13,
   :dewpoint_f 55,
   :feelslike_string "75.0 F (23.9 C)",
   :visibility_km "16.1",
   :icon "clear",
   :nowcast "",
   :pressure_trend "-",
   :temp_c 23.9,
   :wind_gust_mph "1.2",
   :precip_1hr_metric " 0",
   :solarradiation "--",
   :precip_1hr_in "-999.00",
   :heat_index_string "NA",
   :pressure_mb "1012",
   :station_id "KCABEVER22",
   :temperature_string "75.0 F (23.9 C)",
   :observation_time "Last Updated on April 26, 10:32 AM PDT",
   :history_url
   "http://www.wunderground.com/weatherstation/WXDailyHistory.asp?ID=KCABEVER22",
   :observation_time_rfc822 "Wed, 26 Apr 2017 10:32:25 -0700",
   :heat_index_f "NA",
   :local_tz_short "PDT",
   :wind_mph 0.6,
   :local_epoch "1493228336",
   :UV "6",
   :estimated {},
   :feelslike_c "23.9",
   :weather "Clear",
   :display_location
   {:elevation "168.9",
    :full "Beverly Hills, CA",
    :city "Beverly Hills",
    :longitude "-118.41000366",
    :state "CA",
    :magic "1",
    :state_name "California",
    :zip "90210",
    :latitude "34.09999847",
    :country_iso3166 "US",
    :country "US",
    :wmo "99999"},
   :image
   {:url "http://icons.wxug.com/graphics/wu2/logo_130x80.png",
    :title "Weather Underground",
    :link "http://www.wunderground.com"},
   :observation_location
   {:full "Monte Cielo Drive, Beverly Hills, California",
    :city "Monte Cielo Drive, Beverly Hills",
    :state "California",
    :country "US",
    :country_iso3166 "US",
    :latitude "34.099072",
    :longitude "-118.408760",
    :elevation "866 ft"},
   :local_tz_offset "-0700",
   :visibility_mi "10.0",
   :precip_today_metric "0",
   :icon_url "http://icons.wxug.com/i/c/k/clear.gif",
   :wind_kph 1.0,
   :ob_url
   "http://www.wunderground.com/cgi-bin/findweather/getForecast?query=34.099072,-118.408760",
   :wind_dir "West",
   :wind_gust_kph "1.9",
   :precip_today_string "0.00 in (0 mm)",
   :wind_string "Calm",
   :windchill_string "NA",
   :observation_epoch "1493227945"})

(deftest present-parsing
  (testing "present conditions become a simple display string"
    (is (= "Cr75/W1" (pres->str present-data)))))

(def hourly-data
  '({:wdir {:dir "NW", :degrees "310"},
     :windchill {:english "-9999", :metric "-9999"},
     :temp {:english "77", :metric "25"},
     :wspd {:english "10", :metric "16"},
     :pop "0",
     :dewpoint {:english "45", :metric "7"},
     :icon "clear",
     :qpf {:english "0.0", :metric "0"},
     :mslp {:english "29.88", :metric "1012"},
     :feelslike {:english "77", :metric "25"},
     :humidity "33",
     :snow {:english "0.0", :metric "0"},
     :fctcode "1",
     :uvi "9",
     :FCTTIME
     {:weekday_name_abbrev "Wed",
      :epoch "1493233200",
      :min "00",
      :hour "12",
      :yday "115",
      :age "",
      :weekday_name_night_unlang "Wednesday Night",
      :mon_padded "04",
      :hour_padded "12",
      :tz "",
      :isdst "1",
      :civil "12:00 PM",
      :ampm "PM",
      :weekday_name "Wednesday",
      :min_unpadded "0",
      :year "2017",
      :UTCDATE "",
      :weekday_name_night "Wednesday Night",
      :sec "0",
      :month_name_abbrev "Apr",
      :month_name "April",
      :mon "4",
      :pretty "12:00 PM PDT on April 26, 2017",
      :mon_abbrev "Apr",
      :mday_padded "26",
      :weekday_name_unlang "Wednesday",
      :mday "26"},
     :condition "Clear",
     :wx "Sunny",
     :icon_url "http://icons.wxug.com/i/c/k/clear.gif",
     :heatindex {:english "-9999", :metric "-9999"},
     :sky "18"}
    {:wdir {:dir "NNW", :degrees "340"},
     :windchill {:english "-9999", :metric "-9999"},
     :temp {:english "80", :metric "27"},
     :wspd {:english "12", :metric "19"},
     :pop "0",
     :dewpoint {:english "47", :metric "8"},
     :icon "clear",
     :qpf {:english "0.0", :metric "0"},
     :mslp {:english "29.82", :metric "1010"},
     :feelslike {:english "80", :metric "27"},
     :humidity "32",
     :snow {:english "0.0", :metric "0"},
     :fctcode "1",
     :uvi "6",
     :FCTTIME
     {:weekday_name_abbrev "Wed",
      :epoch "1493244000",
      :min "00",
      :hour "15",
      :yday "115",
      :age "",
      :weekday_name_night_unlang "Wednesday Night",
      :mon_padded "04",
      :hour_padded "15",
      :tz "",
      :isdst "1",
      :civil "3:00 PM",
      :ampm "PM",
      :weekday_name "Wednesday",
      :min_unpadded "0",
      :year "2017",
      :UTCDATE "",
      :weekday_name_night "Wednesday Night",
      :sec "0",
      :month_name_abbrev "Apr",
      :month_name "April",
      :mon "4",
      :pretty "3:00 PM PDT on April 26, 2017",
      :mon_abbrev "Apr",
      :mday_padded "26",
      :weekday_name_unlang "Wednesday",
      :mday "26"},
     :condition "Clear",
     :wx "Mostly Sunny",
     :icon_url "http://icons.wxug.com/i/c/k/clear.gif",
     :heatindex {:english "-9999", :metric "-9999"},
     :sky "20"}
    {:wdir {:dir "N", :degrees "354"},
     :windchill {:english "-9999", :metric "-9999"},
     :temp {:english "76", :metric "24"},
     :wspd {:english "14", :metric "23"},
     :pop "0",
     :dewpoint {:english "48", :metric "9"},
     :icon "partlycloudy",
     :qpf {:english "0.0", :metric "0"},
     :mslp {:english "29.8", :metric "1009"},
     :feelslike {:english "76", :metric "24"},
     :humidity "37",
     :snow {:english "0.0", :metric "0"},
     :fctcode "2",
     :uvi "1",
     :FCTTIME
     {:weekday_name_abbrev "Wed",
      :epoch "1493254800",
      :min "00",
      :hour "18",
      :yday "115",
      :age "",
      :weekday_name_night_unlang "Wednesday Night",
      :mon_padded "04",
      :hour_padded "18",
      :tz "",
      :isdst "1",
      :civil "6:00 PM",
      :ampm "PM",
      :weekday_name "Wednesday",
      :min_unpadded "0",
      :year "2017",
      :UTCDATE "",
      :weekday_name_night "Wednesday Night",
      :sec "0",
      :month_name_abbrev "Apr",
      :month_name "April",
      :mon "4",
      :pretty "6:00 PM PDT on April 26, 2017",
      :mon_abbrev "Apr",
      :mday_padded "26",
      :weekday_name_unlang "Wednesday",
      :mday "26"},
     :condition "Partly Cloudy",
     :wx "Partly Cloudy",
     :icon_url "http://icons.wxug.com/i/c/k/partlycloudy.gif",
     :heatindex {:english "-9999", :metric "-9999"},
     :sky "39"}
    {:wdir {:dir "N", :degrees "352"},
     :windchill {:english "-9999", :metric "-9999"},
     :temp {:english "71", :metric "22"},
     :wspd {:english "12", :metric "19"},
     :pop "0",
     :dewpoint {:english "50", :metric "10"},
     :icon "partlycloudy",
     :qpf {:english "0.0", :metric "0"},
     :mslp {:english "29.83", :metric "1010"},
     :feelslike {:english "71", :metric "22"},
     :humidity "48",
     :snow {:english "0.0", :metric "0"},
     :fctcode "2",
     :uvi "0",
     :FCTTIME
     {:weekday_name_abbrev "Wed",
      :epoch "1493265600",
      :min "00",
      :hour "21",
      :yday "115",
      :age "",
      :weekday_name_night_unlang "Wednesday Night",
      :mon_padded "04",
      :hour_padded "21",
      :tz "",
      :isdst "1",
      :civil "9:00 PM",
      :ampm "PM",
      :weekday_name "Wednesday",
      :min_unpadded "0",
      :year "2017",
      :UTCDATE "",
      :weekday_name_night "Wednesday Night",
      :sec "0",
      :month_name_abbrev "Apr",
      :month_name "April",
      :mon "4",
      :pretty "9:00 PM PDT on April 26, 2017",
      :mon_abbrev "Apr",
      :mday_padded "26",
      :weekday_name_unlang "Wednesday",
      :mday "26"},
     :condition "Partly Cloudy",
     :wx "Partly Cloudy",
     :icon_url "http://icons.wxug.com/i/c/k/nt_partlycloudy.gif",
     :heatindex {:english "-9999", :metric "-9999"},
     :sky "30"}))

(deftest hourly-parsing
  (testing "hourly conditions become a simple display string"
    (is (= "Cr77 Cr80 PC76 PC71 " (hourly->str hourly-data)))))

(def daily-data
  '({:date
     {:epoch "1493258400",
      :min "00",
      :day 26,
      :hour 19,
      :yday 115,
      :monthname_short "Apr",
      :isdst "1",
      :month 4,
      :ampm "PM",
      :tz_short "PDT",
      :tz_long "America/Los_Angeles",
      :monthname "April",
      :year 2017,
      :weekday "Wednesday",
      :sec 0,
      :pretty "7:00 PM PDT on April 26, 2017",
      :weekday_short "Wed"},
     :snow_allday {:in 0.0, :cm 0.0},
     :qpf_day {:in 0.0, :mm 0},
     :pop 0,
     :qpf_allday {:in 0.0, :mm 0},
     :avehumidity 33,
     :minhumidity 0,
     :snow_day {:in 0.0, :cm 0.0},
     :icon "clear",
     :high {:fahrenheit "82", :celsius "28"},
     :skyicon "",
     :snow_night {:in 0.0, :cm 0.0},
     :conditions "Clear",
     :avewind {:mph 13, :kph 21, :dir "NNW", :degrees 329},
     :maxhumidity 0,
     :low {:fahrenheit "61", :celsius "16"},
     :period 1,
     :qpf_night {:in 0.0, :mm 0},
     :icon_url "http://icons.wxug.com/i/c/k/clear.gif",
     :maxwind {:mph 20, :kph 32, :dir "NNW", :degrees 329}}
    {:date
     {:epoch "1493344800",
      :min "00",
      :day 27,
      :hour 19,
      :yday 116,
      :monthname_short "Apr",
      :isdst "1",
      :month 4,
      :ampm "PM",
      :tz_short "PDT",
      :tz_long "America/Los_Angeles",
      :monthname "April",
      :year 2017,
      :weekday "Thursday",
      :sec 0,
      :pretty "7:00 PM PDT on April 27, 2017",
      :weekday_short "Thu"},
     :snow_allday {:in 0.0, :cm 0.0},
     :qpf_day {:in 0.0, :mm 0},
     :pop 0,
     :qpf_allday {:in 0.0, :mm 0},
     :avehumidity 35,
     :minhumidity 0,
     :snow_day {:in 0.0, :cm 0.0},
     :icon "partlycloudy",
     :high {:fahrenheit "81", :celsius "27"},
     :skyicon "",
     :snow_night {:in 0.0, :cm 0.0},
     :conditions "Partly Cloudy",
     :avewind {:mph 19, :kph 31, :dir "NNW", :degrees 342},
     :maxhumidity 0,
     :low {:fahrenheit "60", :celsius "16"},
     :period 2,
     :qpf_night {:in 0.0, :mm 0},
     :icon_url "http://icons.wxug.com/i/c/k/partlycloudy.gif",
     :maxwind {:mph 25, :kph 40, :dir "NNW", :degrees 342}}
    {:date
     {:epoch "1493431200",
      :min "00",
      :day 28,
      :hour 19,
      :yday 117,
      :monthname_short "Apr",
      :isdst "1",
      :month 4,
      :ampm "PM",
      :tz_short "PDT",
      :tz_long "America/Los_Angeles",
      :monthname "April",
      :year 2017,
      :weekday "Friday",
      :sec 0,
      :pretty "7:00 PM PDT on April 28, 2017",
      :weekday_short "Fri"},
     :snow_allday {:in 0.0, :cm 0.0},
     :qpf_day {:in 0.0, :mm 0},
     :pop 0,
     :qpf_allday {:in 0.0, :mm 0},
     :avehumidity 24,
     :minhumidity 0,
     :snow_day {:in 0.0, :cm 0.0},
     :icon "clear",
     :high {:fahrenheit "84", :celsius "29"},
     :skyicon "",
     :snow_night {:in 0.0, :cm 0.0},
     :conditions "Clear",
     :avewind {:mph 17, :kph 27, :dir "N", :degrees 350},
     :maxhumidity 0,
     :low {:fahrenheit "60", :celsius "16"},
     :period 3,
     :qpf_night {:in 0.0, :mm 0},
     :icon_url "http://icons.wxug.com/i/c/k/clear.gif",
     :maxwind {:mph 25, :kph 40, :dir "N", :degrees 350}}
    {:date
     {:epoch "1493517600",
      :min "00",
      :day 29,
      :hour 19,
      :yday 118,
      :monthname_short "Apr",
      :isdst "1",
      :month 4,
      :ampm "PM",
      :tz_short "PDT",
      :tz_long "America/Los_Angeles",
      :monthname "April",
      :year 2017,
      :weekday "Saturday",
      :sec 0,
      :pretty "7:00 PM PDT on April 29, 2017",
      :weekday_short "Sat"},
     :snow_allday {:in 0.0, :cm 0.0},
     :qpf_day {:in 0.0, :mm 0},
     :pop 0,
     :qpf_allday {:in 0.0, :mm 0},
     :avehumidity 14,
     :minhumidity 0,
     :snow_day {:in 0.0, :cm 0.0},
     :icon "clear",
     :high {:fahrenheit "84", :celsius "29"},
     :skyicon "",
     :snow_night {:in 0.0, :cm 0.0},
     :conditions "Clear",
     :avewind {:mph 8, :kph 13, :dir "ESE", :degrees 123},
     :maxhumidity 0,
     :low {:fahrenheit "58", :celsius "14"},
     :period 4,
     :qpf_night {:in 0.0, :mm 0},
     :icon_url "http://icons.wxug.com/i/c/k/clear.gif",
     :maxwind {:mph 10, :kph 16, :dir "ESE", :degrees 123}}
    {:date
     {:epoch "1493604000",
      :min "00",
      :day 30,
      :hour 19,
      :yday 119,
      :monthname_short "Apr",
      :isdst "1",
      :month 4,
      :ampm "PM",
      :tz_short "PDT",
      :tz_long "America/Los_Angeles",
      :monthname "April",
      :year 2017,
      :weekday "Sunday",
      :sec 0,
      :pretty "7:00 PM PDT on April 30, 2017",
      :weekday_short "Sun"},
     :snow_allday {:in 0.0, :cm 0.0},
     :qpf_day {:in 0.0, :mm 0},
     :pop 0,
     :qpf_allday {:in 0.0, :mm 0},
     :avehumidity 13,
     :minhumidity 0,
     :snow_day {:in 0.0, :cm 0.0},
     :icon "clear",
     :high {:fahrenheit "82", :celsius "28"},
     :skyicon "",
     :snow_night {:in 0.0, :cm 0.0},
     :conditions "Clear",
     :avewind {:mph 11, :kph 18, :dir "SSW", :degrees 192},
     :maxhumidity 0,
     :low {:fahrenheit "59", :celsius "15"},
     :period 5,
     :qpf_night {:in 0.0, :mm 0},
     :icon_url "http://icons.wxug.com/i/c/k/clear.gif",
     :maxwind {:mph 15, :kph 24, :dir "SSW", :degrees 192}}
    {:date
     {:epoch "1493690400",
      :min "00",
      :day 1,
      :hour 19,
      :yday 120,
      :monthname_short "May",
      :isdst "1",
      :month 5,
      :ampm "PM",
      :tz_short "PDT",
      :tz_long "America/Los_Angeles",
      :monthname "May",
      :year 2017,
      :weekday "Monday",
      :sec 0,
      :pretty "7:00 PM PDT on May 01, 2017",
      :weekday_short "Mon"},
     :snow_allday {:in 0.0, :cm 0.0},
     :qpf_day {:in 0.0, :mm 0},
     :pop 0,
     :qpf_allday {:in 0.0, :mm 0},
     :avehumidity 31,
     :minhumidity 0,
     :snow_day {:in 0.0, :cm 0.0},
     :icon "clear",
     :high {:fahrenheit "81", :celsius "27"},
     :skyicon "",
     :snow_night {:in 0.0, :cm 0.0},
     :conditions "Clear",
     :avewind {:mph 8, :kph 13, :dir "S", :degrees 182},
     :maxhumidity 0,
     :low {:fahrenheit "60", :celsius "16"},
     :period 6,
     :qpf_night {:in 0.0, :mm 0},
     :icon_url "http://icons.wxug.com/i/c/k/clear.gif",
     :maxwind {:mph 10, :kph 16, :dir "S", :degrees 182}}
    {:date
     {:epoch "1493776800",
      :min "00",
      :day 2,
      :hour 19,
      :yday 121,
      :monthname_short "May",
      :isdst "1",
      :month 5,
      :ampm "PM",
      :tz_short "PDT",
      :tz_long "America/Los_Angeles",
      :monthname "May",
      :year 2017,
      :weekday "Tuesday",
      :sec 0,
      :pretty "7:00 PM PDT on May 02, 2017",
      :weekday_short "Tue"},
     :snow_allday {:in 0.0, :cm 0.0},
     :qpf_day {:in 0.0, :mm 0},
     :pop 0,
     :qpf_allday {:in 0.0, :mm 0},
     :avehumidity 38,
     :minhumidity 0,
     :snow_day {:in 0.0, :cm 0.0},
     :icon "clear",
     :high {:fahrenheit "82", :celsius "28"},
     :skyicon "",
     :snow_night {:in 0.0, :cm 0.0},
     :conditions "Clear",
     :avewind {:mph 7, :kph 11, :dir "S", :degrees 171},
     :maxhumidity 0,
     :low {:fahrenheit "60", :celsius "16"},
     :period 7,
     :qpf_night {:in 0.0, :mm 0},
     :icon_url "http://icons.wxug.com/i/c/k/clear.gif",
     :maxwind {:mph 10, :kph 16, :dir "S", :degrees 171}})
  )

(deftest daily-parsing
  (testing "daily conditions become a simple display string"
    (is (= "Cr6182 PC6081 Cr6084 Cr5884 Cr5982 Cr6081 Cr6082"
           (forecast->str daily-data)))))


src/wunder/core.clj
(ns wunder.core
  "wunderbar — current/forecasted weather in i3status 1-line format

  Dependencies: Weather Underground API key"
  (:gen-class)
  (:require [clojure.data.json :as json]
            [clojure.string :as str]
            [environ.core :refer [env]]
            [org.httpkit.client :as http]
            ;; [taoensso.timbre :as timbre :refer [log info debug warn error]]
            [environ.core :refer [env]]))

;; TODO: move defs to defns to avoid work during compile
;; TODO: auto-detect location
;; TODO: standardize on env, not getenv


;; Wunderground API endpoint
(def wapi "http://api.wunderground.com/api/")

;; Wunderground API key every user/dev must have personally
(def wkey (System/getenv "WUNDERKEY"))

(defn debug [& msgs]
  (when (env :wunderdebug)
    (binding [*out* *err*] (apply println msgs))))

(defn abort [& msgs]
  (binding [*out* *err*] (apply println msgs))
  (System/exit 1))

(defn check-key []
  (when-not wkey
    (abort "Must export WUNDERKEY env var."
           "\nSign up for a free key:"
           "\nhttps://www.wunderground.com/weather/api/d/pricing.html")))

;; Default to Beverly Hills where it's sure to be beautiful!
(def default-loc "90210")
(def loc (env :wunderloc default-loc))
;; BUG: what if they really are in 90210??
(when (= loc default-loc)
  (binding [*out* *err*]
    (println "WARNING: Defaulting to" loc "for demo purposes.")))

;; For request query: determine if zip code or personal weather station
(def geoq (if (re-matches #"^\d{5}$" loc)
            ;; Ex: Beaverton zip code: 97005
            ;; Not sure what the 1.99999 junk is for; seems necessary for zips.
            (str "/q/zmw:" loc ".1.99999.json")
            ;; Ex: Beaverton personal weather station (pws): KORBEAVE74
            (str "/q/pws:" loc ".json")))

(defn check-loc []
  (when-not loc
    (abort "Must export WUNDERLOC env var for zip code or nearest station."
           "\nFind a station:"
           "\nhttps://www.wunderground.com/weatherstation/ListStations.asp")))

;; Optionally show PWS location in bar
(def pws-display
  (if (env :wundershowloc)
    (str " (" loc ")")
    ""))

;; (def wunits (or (System/getenv "WUNDERUNITS") :fahrenheit))
;; NOTE: not error checking this user input
(def wunits (or (System/getenv "WUNDERUNITS") "f"))
;; Pretty lame inconsistency of units across calls
(def fc-units   (if (= wunits "f") :fahrenheit :celcius))
(def hrly-units (if (= wunits "f") :english :metric))
(def pres-units (if (= wunits "f") :temp_f :temp_c))

;; Ex: "http://api.wunderground.com/api/abc123/forecast10day/q/zmw:97005.1.99999.json"
(defn mkurl [dur] (str wapi wkey "/" dur geoq))
(def present-url  (mkurl "conditions"))
(def hourly-url   (mkurl "hourly"))
(def forecast-url (mkurl "forecast10day"))
(def ndays 7)  ; number of forecast days

;; Wunderground status codes, based on icon sets.
;; https://www.wunderground.com/weather/api/d/docs?d=resources/icon-sets
;; This is a cute little abbreviation table for compressed display.
;; Surprising how many few codes wunderground has compared to
;; openweathermap.
(def conds {"chanceflurries" "CF"
            "chancerain"     "CR"
            "chancesleet"    "CH"
            "chancesnow"     "CS"
            "chancetstorms"  "CT"
            "clear"          "Cr"
            "cloudy"         "Cd"
            "flurries"       "Fl"
            "fog"            "Fg"
            "hazy"           "Hz"
            "mostlycloudy"   "MC"
            "mostlysunny"    "MS"
            "partlycloudy"   "PC"
            "partlysunny"    "PS"
            "rain"           "Rn"
            "sleet"          "Hl"
            "snow"           "Sw"
            "sunny"          "Sy"
            "tstorms"        "Th"})

;; Could grab all three URLs in parallel, but not worth it yet.
;; Really should check for full payload containing valid JSON, and
;; retry if not.
(defn wunder-request
  "Request the weather data json for a wunderground url.
  Called a few times to get present, hourly, and daily conditions.

  NOTE: Not part of test suite since heavy network calls."
  [url]
  (debug "requesting url:" url)
  (let [{:keys [status headers body error] :as resp}
        @(http/get url)]
    (if error
      ;; Silly API does not give errors, but shows good resp with
      ;; error in text, so not really doing anything with problem case.
      (println "error: " error)
      ;; Read the JSON and convert to hashmap.
      (json/read-json body true))))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Present conditions

(defn select-present
  "Grab the blob of present-time observations."
  []
  (get-in (wunder-request present-url) [:current_observation]))

(defn pres->str
  "Massage present conditions metrics to desired units as strings.

  Input: large map of present snapshot
  Output: \"Cr79/W1\""
  [pc] ; present conditions
  (let [cnd    (get conds (:icon pc))
        temp   (str (Math/round (pres-units pc)))
        barom  (str "B" (:pressure_in pc))
        wind   (str "W" (Math/round (:wind_mph pc)))
        precip (str "P" (:precip_1hr_in pc))]
    ;; Not using all, but doesn't hurt to have captured them
    (str cnd temp "/" wind)))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Hourly forecast for today

(defn select-hours
  "Grab pieces of hourly blob; every third hour."
  []
  (take 4 (take-nth 3 (:hourly_forecast
                       (wunder-request hourly-url)))))

;; (hourly->str (select-hours))
(defn hourly->str
  "Massage hourly conditions into a string.

  Input: large map of hourly data
  Output: \"CR54 CR54 CR53 Rn50 \""
  [hc] ; hourly conditions
  (map (juxt :icon :temp :wdir) hc)
  (let [cnds (map #(get conds (:icon %)) hc)
        temp (map #(get-in % [:temp hrly-units]) hc)
        wdir (map #(get-in % [:wdir :dir]) hc)]
    (apply str (interleave cnds temp (repeat " ")))))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Daily forecasting for this week

(defn select-days
  "Request/transform into list of forecast days data"
  []
  (take ndays (get-in (wunder-request forecast-url)
                      [:forecast :simpleforecast :forecastday])))

;; '(("CR" "Rn" "Rn" "Rn") ("45" "45" "47" "47") ("67" "53" "54" "53"))
(defn lohis
  "Generate seq of conds, los, his from `days` (large structure)"
  [days]
  (let [lhs  (for [lohi [:low :high]]
               (map #(get-in % [lohi fc-units]) days))
        cnds (map #(get conds (:icon %)) days)]
    (conj lhs cnds)))

(defn forecast->str
  "Massage the whole week into a string.

  Input: large map of week forecast data
  Output: \"CR4567 Rn4553 Rn4754 Rn4753\""
  [days]
  (let [ds (map #(apply str %)
                (partition 3 (apply interleave (lohis days))))]
    (str/join " " ds)))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Main

(defn -main
  "Put present, hourly, daily together into output for status bar

  Output: \"Cl52/W2/P-999.00 | CR51 Rn48 Rn47 Rn46 | Rn4556 Rn4353 Rn4852 Rn4252 CR4054 CR3958 PC4463\"
  "
  [& args]
  (check-key)
  (debug "fetching weather for" loc)
  (let [pres (pres->str (select-present))
        hrly (hourly->str (select-hours))
        dyly (forecast->str (select-days))]
    (spit "/tmp/wunderbar.txt" (str pres pws-display " | " hrly "| " dyly))))

;; (-main) test run
